File: js/browser_event_demo.js
window.addEventListener('load', e => {
  const canvas = document.createElement('canvas');
  const ctx = canvas.getContext('2d');
  let drawing = false;
  canvas.style.border = '2px solid purple';
  canvas.addEventListener('mousedown', e => {
    drawing = true;
    ctx.moveTo(e.clientX, e.clientY);
  });
  canvas.addEventListener('mousemove', e => {
    if (drawing) {
      ctx.lineTo(e.clientX, e.clientY);
      ctx.stroke();
    }
  });
  const stopDrawing = e => {drawing = false};
  canvas.addEventListener('mouseup', stopDrawing);
  canvas.addEventListener('mouseout', stopDrawing);
  document.body.appendChild(canvas);
});

File: dart/browser_event_demo.dart
import 'dart:html';

void main() {
  var canvas = new Element.canvas();
  var ctx = canvas.context2D;
  var drawing = false;
  canvas.style.border = '2px solid green';
  canvas.onMouseDown.listen((e) {
    drawing = true;
    ctx.moveTo(e.client.x, e.client.y);
  });
  canvas.onMouseMove.listen((e) {
    if (drawing) {
      ctx.lineTo(e.client.x, e.client.y);
      ctx.stroke();
    }
  });
  var stop = (e) => drawing = false;
  canvas
      ..onMouseOut.listen(stop)
      ..onMouseUp.listen(stop);

  document.body.children.add(canvas);
}
